**Catatan**: File PDF dalam sesi ini dapat dilihat pada github berikut:
https://github.com/Sardiirfan27/h8_llm_genai/blob/main/docs_telco_v2.pdf

# Membaca Dokumen PDF

Sebelum mencoba implementasi chunking, terlebih dahulu kita akan membaca dokumen PDF yang akan kita chunking. Ada macam-macam cara untuk membaca dokumen PDF di Python. Kita akan menggunakan library:
- pypdf: Ringan dan mudah digunakan. Cocok untuk PDF berbasis teks (bukan scan), tetapi hasil ekstraksi kadang berantakan dan kurang bagus untuk layout kompleks (tabel, kolom).
- PyMuPDF: Menyediakan fitur tambahan seperti ekstraksi tabel dan cukup baik untuk layout yang kompleks.
- pymupdf4llm: Menyediakan ekstraksi teks dalam format Markdown yang lebih rapi dan mudah dibaca oleh AI.

Versi Library:
- "pypdf>=6.7.3"
- "pymupdf>=1.27.1"
- "pymupdf4llm>=0.3.4"

In [ ]:
# deklarasi variabel untuk path PDF
pdf_path = "docs_telco_v2.pdf"

## pypdf

Sekarang, kita akan membaca dokumen PDF menggunakan library pypdf untuk melihat hasil ekstraksi teksnya.

In [ ]:
from pypdf import PdfReader
reader = PdfReader(pdf_path)
page = reader.pages[0] # ambil halaman pertama
print(page.extract_text())

PT  Nusantara  Telekomunikasi  Digital  
(NTD)
 
 
1.  Profil  Perusahaan  
PT  Nusantara  Telekomunikasi  Digital  (NTD)  adalah  perusahaan  telekomunikasi  nasional  
yang
 
berdiri
 
pada
 
tahun
 
2008
 
dan
 
berfokus
 
pada
 
penyediaan
 
layanan
 
jaringan
 
seluler,
 
internet
 
broadband,
 
serta
 
solusi
 
digital
 
enterprise.
 
Perusahaan
 
ini
 
beroperasi
 
di
 
lebih
 
dari
 
150
 
kota
 
di
 
Indonesia
 
dan
 
memiliki
 
lebih
 
dari
 
35
 
juta
 
pelanggan
 
aktif.
 
NTD  menyediakan  layanan  4G  LTE  dan  telah  mulai  melakukan  ekspansi  jaringan  5G  secara  
bertahap
 
di
 
kota-kota
 
besar
 
seperti
 
Jakarta,
 
Surabaya,
 
Bandung,
 
dan
 
Medan.
 
Selain
 
layanan
 
ritel,
 
NTD
 
juga
 
memiliki
 
divisi
 
Business
 
&
 
Enterprise
 
yang
 
melayani
 
kebutuhan
 
konektivitas
 
korporasi,
 
data
 
center,
 
dan
 
solusi
 
cloud.
 
Ringkasan  Perusahaan  
Kategori  Detail  
Tahun  Berdiri  2008  
Jumlah  Pelanggan  
35+  juta  
Cakupan  Kota  150+  kota  
Te

Hasilnya sangat berantakan, selanjutnya kita akan menggunakan mode ekstraksi "layout" untuk menghasilkan teks yang sesuai dengan tata letak yang dihasilkan dalam sumber PDF.

In [ ]:
# Ekstrak teks dalam format lebar tetap yang sesuai dengan tata letak yang dihasilkan dalam sumber PDF
result_text = page.extract_text(extraction_mode="layout")
print('total karakter:', len(result_text))
print(result_text)

total karakter: 973
PT Nusantara Telekomunikasi Digital
(NTD)



1. Profil Perusahaan

PT Nusantara Telekomunikasi Digital (NTD) adalah perusahaan telekomunikasi nasional
yang berdiri pada tahun 2008 dan berfokus pada penyediaan layanan jaringan seluler,
internet broadband, serta solusi digital enterprise. Perusahaan ini beroperasi di lebih dari
150 kota di Indonesia dan memiliki lebih dari 35 juta pelanggan aktif.

NTD menyediakan layanan 4G LTE dan telah mulai melakukan ekspansi jaringan 5G secara
bertahap di kota-kota besar seperti Jakarta, Surabaya, Bandung, dan Medan. Selain
layanan ritel, NTD juga memiliki divisi Business & Enterprise yang melayani kebutuhan
konektivitas korporasi, data center, dan solusi cloud.

Ringkasan Perusahaan


      Kategori                 Detail



 Tahun Berdiri          2008



 Jumlah                 35+ juta
 Pelanggan



 Cakupan Kota           150+ kota



 Teknologi Utama        4G LTE, 5G



 Segmen Bisnis          Retail & Enterprise





Hasilnya lebih rapi dan masih memiliki spasi vertikal berlebih yang dapat memperluas ukuran teks. Untuk menghilangkannya, kita dapat mengatur opsi "layout_mode_space_vertically" menjadi False.

In [ ]:
# Ekstrak teks dengan mempertahankan posisi horizontal tanpa spasi vertikal berlebih
# Menghapus whitespace vertikal
result_text = page.extract_text(extraction_mode="layout", layout_mode_space_vertically=False)
print('total karakter:', len(result_text))
print(result_text)

total karakter: 947
PT Nusantara Telekomunikasi Digital
(NTD)
1. Profil Perusahaan
PT Nusantara Telekomunikasi Digital (NTD) adalah perusahaan telekomunikasi nasional
yang berdiri pada tahun 2008 dan berfokus pada penyediaan layanan jaringan seluler,
internet broadband, serta solusi digital enterprise. Perusahaan ini beroperasi di lebih dari
150 kota di Indonesia dan memiliki lebih dari 35 juta pelanggan aktif.
NTD menyediakan layanan 4G LTE dan telah mulai melakukan ekspansi jaringan 5G secara
bertahap di kota-kota besar seperti Jakarta, Surabaya, Bandung, dan Medan. Selain
layanan ritel, NTD juga memiliki divisi Business & Enterprise yang melayani kebutuhan
konektivitas korporasi, data center, dan solusi cloud.
Ringkasan Perusahaan
      Kategori                 Detail
 Tahun Berdiri          2008
 Jumlah                 35+ juta
 Pelanggan
 Cakupan Kota           150+ kota
 Teknologi Utama        4G LTE, 5G
 Segmen Bisnis          Retail & Enterprise


Dari hasil di atas, ukuran teks lebih sedikit namun tetap rapi.

Selanjutnya, kita juga dapat menambahkan pengaturan untuk mengatur spasi horizontal.

In [ ]:
# adjust horizontal spacing
result_text = page.extract_text(extraction_mode="layout", layout_mode_scale_weight=1, layout_mode_space_vertically=False)
print('total karakter:', len(result_text))
print(result_text)

total karakter: 915
PT Nusantara Telekomunikasi Digital
(NTD)
1. Profil Perusahaan
PT Nusantara Telekomunikasi Digital (NTD) adalah perusahaan telekomunikasi nasional
yang berdiri pada tahun 2008 dan berfokus pada penyediaan layanan jaringan seluler,
internet broadband, serta solusi digital enterprise. Perusahaan ini beroperasi di lebih dari
150 kota di Indonesia dan memiliki lebih dari 35 juta pelanggan aktif.
NTD menyediakan layanan 4G LTE dan telah mulai melakukan ekspansi jaringan 5G secara
bertahap di kota-kota besar seperti Jakarta, Surabaya, Bandung, dan Medan. Selain
layanan ritel, NTD juga memiliki divisi Business & Enterprise yang melayani kebutuhan
konektivitas korporasi, data center, dan solusi cloud.
Ringkasan Perusahaan
     Kategori            Detail
Tahun Berdiri      2008
Jumlah             35+ juta
Pelanggan
Cakupan Kota       150+ kota
Teknologi Utama    4G LTE, 5G
Segmen Bisnis      Retail & Enterprise


Sekilas hasil tersebut mirip dengan hasil sebelumnya namun total dari karakternya lebih sedikit.

Salah satu keterbatasan pypdf adalah tidak adanya kemampuan khusus untuk mendeteksi dan merekonstruksi tabel sebagai struktur terpisah. Saat mengekstrak dokumen yang mengandung tabel, hasil yang diperoleh biasanya berupa teks biasa tanpa pemisahan kolom yang jelas. Akibatnya, informasi yang seharusnya berbentuk baris dan kolom akan tergabung menjadi teks linear.

## PyMuPDF

Selanjutnya kita akan coba menggunakan PyMuPDF (fitz) yang dapat membaca tabel dan mengakses koordinat teks.

In [ ]:
import fitz  # PyMuPDF
from tabulate import tabulate

doc = fitz.open(pdf_path)
doc_output = ""

for page in doc:

    elements = []

    # 1️⃣ Detect tables
    tables = page.find_tables()
    table_bboxes = [table.bbox for table in tables.tables]

    # 2️⃣ Get text blocks
    blocks = page.get_text("blocks")

    for block in blocks:
        x0, y0, x1, y1, text, *_ = block

        # cek apakah block berada dalam area tabel
        inside_table = False
        for tb in table_bboxes:
            tx0, ty0, tx1, ty1 = tb
            if (x0 >= tx0 and x1 <= tx1 and
                y0 >= ty0 and y1 <= ty1):
                inside_table = True
                break

        if not inside_table:
            elements.append({
                "type": "text",
                "top": y0,
                "content": text.strip()
            })

    # 3️⃣ Add tables
    for table in tables.tables:
        elements.append({
            "type": "table",
            "top": table.bbox[1],
            "content": table.extract()
        })

    # 4️⃣ Sort by vertical position
    elements = sorted(elements, key=lambda x: x["top"])

    # 5️⃣ Render to markdown
    for el in elements:
        if el["type"] == "text":
            doc_output += el["content"] + "\n\n"
        else:
            table_data = el["content"]
            headers = table_data[0]
            rows = table_data[1:]
            md_table = tabulate(rows, headers=headers, tablefmt="github")
            doc_output += md_table + "\n\n"

print(doc_output)

PT Nusantara Telekomunikasi Digital 
(NTD)



1. Profil Perusahaan

PT Nusantara Telekomunikasi Digital (NTD) adalah perusahaan telekomunikasi nasional 
yang berdiri pada tahun 2008 dan berfokus pada penyediaan layanan jaringan seluler, 
internet broadband, serta solusi digital enterprise. Perusahaan ini beroperasi di lebih dari 
150 kota di Indonesia dan memiliki lebih dari 35 juta pelanggan aktif.

NTD menyediakan layanan 4G LTE dan telah mulai melakukan ekspansi jaringan 5G secara 
bertahap di kota-kota besar seperti Jakarta, Surabaya, Bandung, dan Medan. Selain 
layanan ritel, NTD juga memiliki divisi Business & Enterprise yang melayani kebutuhan 
konektivitas korporasi, data center, dan solusi cloud.

Ringkasan Perusahaan

| Kategori        | Detail              |
|-----------------|---------------------|
| Tahun Berdiri   | 2008                |
| Jumlah
Pelanggan                 | 35+ juta            |
| Cakupan Kota    | 150+ kota           |
| Teknologi Utama | 4G LTE, 5G     

Hasil di atas tampak lebih rapi dan terdapat tabel yang sesuai dengan tata letak yang dihasilkan dalam sumber PDF. Agar lebih rapi lagi, jika ingin menghapus line break berlebih, maka dapat menggunakan cara sebagai berikut

In [ ]:
import re
doc_output = re.sub(r'\n{3,}', '\n\n', doc_output) # mengganti 3 atau lebih newline dengan 2 newline
doc_output = doc_output.strip()

print('total karakter:', len(doc_output))
print(doc_output)

total karakter: 5800
PT Nusantara Telekomunikasi Digital 
(NTD)

1. Profil Perusahaan

PT Nusantara Telekomunikasi Digital (NTD) adalah perusahaan telekomunikasi nasional 
yang berdiri pada tahun 2008 dan berfokus pada penyediaan layanan jaringan seluler, 
internet broadband, serta solusi digital enterprise. Perusahaan ini beroperasi di lebih dari 
150 kota di Indonesia dan memiliki lebih dari 35 juta pelanggan aktif.

NTD menyediakan layanan 4G LTE dan telah mulai melakukan ekspansi jaringan 5G secara 
bertahap di kota-kota besar seperti Jakarta, Surabaya, Bandung, dan Medan. Selain 
layanan ritel, NTD juga memiliki divisi Business & Enterprise yang melayani kebutuhan 
konektivitas korporasi, data center, dan solusi cloud.

Ringkasan Perusahaan

| Kategori        | Detail              |
|-----------------|---------------------|
| Tahun Berdiri   | 2008                |
| Jumlah
Pelanggan                 | 35+ juta            |
| Cakupan Kota    | 150+ kota           |
| Teknologi Utam

## pymupdf4llm

Selain menggunakan PyMuPDF, kita juga dapat menggunakan library lain seperti pymupdf4llm untuk membaca dokumen PDF dan menghasilkan teks dalam format Markdown. Format ini umumnya lebih rapi dan mudah dibaca oleh AI.

Dokumentasinya dapat dibaca di [sini](https://pymupdf.readthedocs.io/en/latest/pymupdf4llm/).

In [ ]:
import pymupdf4llm
md_text = pymupdf4llm.to_markdown(pdf_path, write_images=True)
md_text = re.sub(r'\n{3,}', '\n\n', md_text) # Ganti semua 3+ newline menjadi hanya 2 newline.
md_text = md_text.strip()

print('total karakter:', len(md_text))
print(md_text)

total karakter: 4503
# **PT Nusantara Telekomunikasi Digital** **(NTD)**

## **1. Profil Perusahaan**

PT Nusantara Telekomunikasi Digital (NTD) adalah perusahaan telekomunikasi nasional
yang berdiri pada tahun 2008 dan berfokus pada penyediaan layanan jaringan seluler,
internet broadband, serta solusi digital enterprise. Perusahaan ini beroperasi di lebih dari
150 kota di Indonesia dan memiliki lebih dari 35 juta pelanggan aktif.

NTD menyediakan layanan 4G LTE dan telah mulai melakukan ekspansi jaringan 5G secara
bertahap di kota-kota besar seperti Jakarta, Surabaya, Bandung, dan Medan. Selain
layanan ritel, NTD juga memiliki divisi Business & Enterprise yang melayani kebutuhan
konektivitas korporasi, data center, dan solusi cloud.

### **Ringkasan Perusahaan**

|Kategori|Detail|
|---|---|
|Tahun Berdiri|2008|
|Jumlah<br>Pelanggan|35+ juta|
|Cakupan Kota|150+ kota|
|Teknologi Utama|4G LTE, 5G|
|Segmen Bisnis|Retail & Enterprise|

## **2. Layanan Konsumen (Retail Services)**

### **a.

# Chunking Strategies

## Fixed Chunking

Pada bagian ini, kita akan mencoba melakukan fixed chungking dengan ukuran chunk sebesar 500 karakter. Kita akan mencoba membuat fungsi custom tanpa menggunakan langchain. Pada fungsi tersebut, kita akan melakukan chunking tanpa overlap.

`Overlap` berarti tiap chunk mundur sebanyak karakter yang ditentukan agar konteks tidak terpotong

In [ ]:
# chunking tanpa overlap dengan ukuran tetap tanpa overlap
def fixed_chunk(text, chunk_size=500):
    chunks = []
    for i in range(0, len(text), chunk_size):
        chunk = text[i:i+chunk_size]
        chunks.append(chunk)
    return chunks


# terapkan fungsi chunking pada teks markdown yang dihasilkan
fixed_chunks = fixed_chunk(md_text, 500)

print("Jumlah chunk:", len(fixed_chunks))
print("Contoh chunk pertama:\n")
print(fixed_chunks[0])

Jumlah chunk: 10
Contoh chunk pertama:

# **PT Nusantara Telekomunikasi Digital** **(NTD)**

## **1. Profil Perusahaan**

PT Nusantara Telekomunikasi Digital (NTD) adalah perusahaan telekomunikasi nasional
yang berdiri pada tahun 2008 dan berfokus pada penyediaan layanan jaringan seluler,
internet broadband, serta solusi digital enterprise. Perusahaan ini beroperasi di lebih dari
150 kota di Indonesia dan memiliki lebih dari 35 juta pelanggan aktif.

NTD menyediakan layanan 4G LTE dan telah mulai melakukan ekspansi jaringan 5G secara



In [ ]:
print("Contoh chunk kedua:\n")
print(fixed_chunks[1])

Contoh chunk kedua:

bertahap di kota-kota besar seperti Jakarta, Surabaya, Bandung, dan Medan. Selain
layanan ritel, NTD juga memiliki divisi Business & Enterprise yang melayani kebutuhan
konektivitas korporasi, data center, dan solusi cloud.

### **Ringkasan Perusahaan**

|Kategori|Detail|
|---|---|
|Tahun Berdiri|2008|
|Jumlah<br>Pelanggan|35+ juta|
|Cakupan Kota|150+ kota|
|Teknologi Utama|4G LTE, 5G|
|Segmen Bisnis|Retail & Enterprise|

## **2. Layanan Konsumen (Retail Services)**

### **a. Paket Prabayar**

Pa


Selanjutnya kita akan mencoba melakukan fixed chunking dengan overlap.

In [ ]:
def fixed_chunk_overlap(text, chunk_size=500, overlap=100):
    chunks = []
    step = chunk_size - overlap

    for i in range(0, len(text), step):
        chunk = text[i:i+chunk_size]
        chunks.append(chunk)

        if i + chunk_size >= len(text):
            break

    return chunks


overlap_chunks = fixed_chunk_overlap(md_text, 500, 100)

print("Jumlah chunk:", len(overlap_chunks))
print("Contoh chunk pertama:\n")
print(overlap_chunks[0])

Jumlah chunk: 12
Contoh chunk pertama:

# **PT Nusantara Telekomunikasi Digital** **(NTD)**

## **1. Profil Perusahaan**

PT Nusantara Telekomunikasi Digital (NTD) adalah perusahaan telekomunikasi nasional
yang berdiri pada tahun 2008 dan berfokus pada penyediaan layanan jaringan seluler,
internet broadband, serta solusi digital enterprise. Perusahaan ini beroperasi di lebih dari
150 kota di Indonesia dan memiliki lebih dari 35 juta pelanggan aktif.

NTD menyediakan layanan 4G LTE dan telah mulai melakukan ekspansi jaringan 5G secara



In [ ]:
print("Contoh chunk kedua:\n")
print(overlap_chunks[1])

Contoh chunk kedua:

anggan aktif.

NTD menyediakan layanan 4G LTE dan telah mulai melakukan ekspansi jaringan 5G secara
bertahap di kota-kota besar seperti Jakarta, Surabaya, Bandung, dan Medan. Selain
layanan ritel, NTD juga memiliki divisi Business & Enterprise yang melayani kebutuhan
konektivitas korporasi, data center, dan solusi cloud.

### **Ringkasan Perusahaan**

|Kategori|Detail|
|---|---|
|Tahun Berdiri|2008|
|Jumlah<br>Pelanggan|35+ juta|
|Cakupan Kota|150+ kota|
|Teknologi Utama|4G LTE, 5G|
|Segmen Bisn


Selanjutnya kita akan coba dengan langchain untuk melakukan text splitter dan chunking.

In [ ]:
from langchain_text_splitters import CharacterTextSplitter

splitter = CharacterTextSplitter(
    chunk_size=500,      # setiap chunk maksimal 500 karakter
    chunk_overlap=100,    # overlap 100 karakter
    separator="\n\n" # ubah separtor menjadi separator="", jika ingin mengabaikan pemisah antar chunk.
)

chunks = splitter.split_text(md_text)

print("Total chunk:", len(chunks))
print("Contoh chunk pertama (total karakter):", len(chunks[0]),'\n')

print(chunks[0])

Total chunk: 12
Contoh chunk pertama (total karakter): 413 

# **PT Nusantara Telekomunikasi Digital** **(NTD)**

## **1. Profil Perusahaan**

PT Nusantara Telekomunikasi Digital (NTD) adalah perusahaan telekomunikasi nasional
yang berdiri pada tahun 2008 dan berfokus pada penyediaan layanan jaringan seluler,
internet broadband, serta solusi digital enterprise. Perusahaan ini beroperasi di lebih dari
150 kota di Indonesia dan memiliki lebih dari 35 juta pelanggan aktif.


CharacterTextSplitter: adalah metode untuk memecah teks berdasarkan jumlah karakter dan kondisi separator. Kekurangan dari teknik ini adalah seringkali memotong di tengah kalimat atau paragraf jika jumlah karakternya habis , sehingga konteks semantik hilang.

Selain CharacterTextSplitter, kita dapat menggunakan RecursiveCharacterSplitter yang umumnya menghasilkan chunk yang lebih rapi dan mengatasi kekurangan dari CharacterTextSplitter, walaupun sebenarnya bisa juga menyesuaikan separatornya agar lebih rapi dan mirip seperti RecursiveChacaterTextSplitter.

In [ ]:
from langchain_text_splitters import RecursiveCharacterTextSplitter
rc_splitter = RecursiveCharacterTextSplitter(
    chunk_size=500,
    chunk_overlap=100,

)
chunks = rc_splitter.split_text(md_text)

print("Total chunk:", len(chunks))
print("Contoh chunk pertama (total karakter):", len(chunks[0]),'\n')
print(chunks[0])

Total chunk: 12
Contoh chunk pertama (total karakter): 413 

# **PT Nusantara Telekomunikasi Digital** **(NTD)**

## **1. Profil Perusahaan**

PT Nusantara Telekomunikasi Digital (NTD) adalah perusahaan telekomunikasi nasional
yang berdiri pada tahun 2008 dan berfokus pada penyediaan layanan jaringan seluler,
internet broadband, serta solusi digital enterprise. Perusahaan ini beroperasi di lebih dari
150 kota di Indonesia dan memiliki lebih dari 35 juta pelanggan aktif.


## Recursive Chunking

Untuk melakukan Recursive Chunking kita dapat gunakan RecursiveCharacterTextSplitter dengan penambahan konfigurasi di separatornya.

Misalkan, kita telah mengetahui markdown kita ada heading dan kita mau menambahkan separator berdasarkan heading selain separator paragraf, newline, dan spasi. Heading utama (H1) dengan `"\n# "`, dan heading kedua (subheading) dengan `"\n## "`.


In [ ]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

# Recursive splitter bisa memanfaatkan heading
splitter = RecursiveCharacterTextSplitter(
    chunk_size=500,
    chunk_overlap=100,
      separators=["\n## ", "\n# ", "\n\n", "\n", " ", ""]
)

chunks = splitter.split_text(md_text)

print("Total chunk:", len(chunks))
print("Contoh chunk pertama (total karakter):", len(chunks[0]),'\n')
print(chunks[0])

Total chunk: 14
Contoh chunk pertama (total karakter): 51 

# **PT Nusantara Telekomunikasi Digital** **(NTD)**


In [ ]:
for i, chunk in enumerate(chunks):
    print(f"Chunk {i} (total karakter):", len(chunk))
    print(chunk)
    print("-" * 50)

Chunk 0 (total karakter): 51
# **PT Nusantara Telekomunikasi Digital** **(NTD)**
--------------------------------------------------
Chunk 1 (total karakter): 360
## **1. Profil Perusahaan**

PT Nusantara Telekomunikasi Digital (NTD) adalah perusahaan telekomunikasi nasional
yang berdiri pada tahun 2008 dan berfokus pada penyediaan layanan jaringan seluler,
internet broadband, serta solusi digital enterprise. Perusahaan ini beroperasi di lebih dari
150 kota di Indonesia dan memiliki lebih dari 35 juta pelanggan aktif.
--------------------------------------------------
Chunk 2 (total karakter): 337
NTD menyediakan layanan 4G LTE dan telah mulai melakukan ekspansi jaringan 5G secara
bertahap di kota-kota besar seperti Jakarta, Surabaya, Bandung, dan Medan. Selain
layanan ritel, NTD juga memiliki divisi Business & Enterprise yang melayani kebutuhan
konektivitas korporasi, data center, dan solusi cloud.

### **Ringkasan Perusahaan**
--------------------------------------------------
Chunk 3

Jika kita ingin fokus hanya pada heading, efeknya adalah beberapa chunk akan memiliki ukuran karakter lebih dari 500 atau bisa diartikan bahwa seperator akan menjadi pertimbangan awal atau utama sebelumnya melihat chunk_size.

In [ ]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

# Recursive splitter bisa memanfaatkan heading
splitter = RecursiveCharacterTextSplitter(
    chunk_size=500,
    chunk_overlap=100,
    separators=["\n## ", "\n# "] # fokus hanya pada heading untuk memisahkan chunk
)

chunks = splitter.split_text(md_text)

print("Total chunk:", len(chunks))
for i, chunk in enumerate(chunks):
    print(f"Chunk {i} (total karakter):", len(chunk))
    print(chunk)
    print("-" * 50)

Total chunk: 10
Chunk 0 (total karakter): 51
# **PT Nusantara Telekomunikasi Digital** **(NTD)**
--------------------------------------------------
Chunk 1 (total karakter): 872

## **1. Profil Perusahaan**

PT Nusantara Telekomunikasi Digital (NTD) adalah perusahaan telekomunikasi nasional
yang berdiri pada tahun 2008 dan berfokus pada penyediaan layanan jaringan seluler,
internet broadband, serta solusi digital enterprise. Perusahaan ini beroperasi di lebih dari
150 kota di Indonesia dan memiliki lebih dari 35 juta pelanggan aktif.

NTD menyediakan layanan 4G LTE dan telah mulai melakukan ekspansi jaringan 5G secara
bertahap di kota-kota besar seperti Jakarta, Surabaya, Bandung, dan Medan. Selain
layanan ritel, NTD juga memiliki divisi Business & Enterprise yang melayani kebutuhan
konektivitas korporasi, data center, dan solusi cloud.

### **Ringkasan Perusahaan**

|Kategori|Detail|
|---|---|
|Tahun Berdiri|2008|
|Jumlah<br>Pelanggan|35+ juta|
|Cakupan Kota|150+ kota|
|Teknologi Utam

Perhatikan pada chunk 1 total karakterknya adalah 872 karena syarat pemisahannya fokus ke H1 dan H2 tanpa melihat new line.

> **Catatan**: Data yang telah di split atau dilakukan chunking nantinya akan kita gunakan pada sesi latihan setelah memperlajari Metadata dan Hybrid Search.

# Metadata & Hybrid Search

`Metadata` adalah suatu data atau informasi tambahan yang dapat membantu untuk menjelaskan, mengelompokkan, atau memberi konteks pada suatu dokumen.

Kegunaan metadata diantaranya yaitu:
- Membuat Search Lebih Presisi karena bisa filter dokumen atau data.
- Mengurangi noise pada hasil pencarian dengan mengeliminasi dokumen yang tidak relevan atau membatasi domain pencarian.
- Membantu dalam segmentasi data.


Selain metadata, kita dapat menggunakan hybrid search untuk melakukan pencarian dengan lebih akurat.
Hybrid search adalah gabungan dua metode pencarian berbeda secara scoring, biasanya:
- Dense retrieval (semantic search)
- Sparse retrieval (keyword search)

Sekarang kita akan mencoba membuat dataset terlebih dahulu sebelum melakukan melakukan pencarian.

In [ ]:
import pandas as pd

data = {
    "id": [1, 2, 3, 4, 5],
    "nama_produk": [
        "Sepatu lari ringan pria",
        "Sepatu casual wanita",
        "Laptop gaming RTX 4060",
        "Laptop kerja tipis ringan",
        "Headphone noise cancelling"
    ],
    "deskripsi": [
        "Sepatu ringan dengan bantalan empuk cocok untuk olahraga dan marathon.",
        "Sepatu casual stylish dan nyaman untuk aktivitas sehari-hari.",
        "Laptop gaming performa tinggi dengan GPU RTX 4060 dan RAM 16GB.",
        "Laptop tipis dan ringan cocok untuk kerja kantoran dan presentasi.",
        "Headphone dengan fitur noise cancelling aktif dan suara premium."
    ],
    "kategori": ["sepatu", "sepatu", "laptop", "laptop", "aksesoris"],
    "brand": ["Nike", "Adidas", "Asus", "Lenovo", "Sony"],
    "harga": [950000, 850000, 18500000, 12500000, 2500000],
    "rating": [4.7, 4.5, 4.8, 4.6, 4.9],
    "stok": [True, True, False, True, True]
}

df = pd.DataFrame(data)
display(df)

,id,nama_produk,deskripsi,kategori,brand,harga,rating,stok
0,1,Sepatu lari ringan pria,Sepatu ringan dengan bantalan empuk cocok untu...,sepatu,Nike,950000,4.7,True
1,2,Sepatu casual wanita,Sepatu casual stylish dan nyaman untuk aktivit...,sepatu,Adidas,850000,4.5,True
2,3,Laptop gaming RTX 4060,Laptop gaming performa tinggi dengan GPU RTX 4...,laptop,Asus,18500000,4.8,False
3,4,Laptop kerja tipis ringan,Laptop tipis dan ringan cocok untuk kerja kant...,laptop,Lenovo,12500000,4.6,True
4,5,Headphone noise cancelling,Headphone dengan fitur noise cancelling aktif ...,aksesoris,Sony,2500000,4.9,True


Untuk melakukan search, kita akan coba menggunakan chromadb dan pinecone sebagai database untuk menyimpan metadata dari data tersebut.

## ChromaDB

In [ ]:
# # delete collection
# client.delete_collection("ecommerce")

In [ ]:
import chromadb
from chromadb.config import Settings
from sentence_transformers import SentenceTransformer

# Inisialisasi model embedding dan ChromaDB client
model = SentenceTransformer("all-MiniLM-L6-v2")
client = chromadb.Client(
    Settings(
        persist_directory="./chroma_db",
        is_persistent=True
    )
)


# Buat collection baru di ChromaDB
collection = client.create_collection("ecommerce",
                                      metadata={"hnsw:space": "cosine"}
                                      )

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Pada konfigurasi diatas, kita akan mencoba menggunakan cosine distance untuk menghitung jarak antar embedding.

Untuk pembuatan dokumen, kita dapat menggabungkan `nama_produk` dengan `deskripsi`. Ingat bahwa dokumen inilah yang nantinya akan di embedding.

In [ ]:
# Gabungkan nama + deskripsi jadi 1 document
documents = [
f"""Nama Produk: {row.nama_produk}.
Deskripsi: {row.deskripsi}
"""
for _, row in df.iterrows()
]

documents

['Nama Produk: Sepatu lari ringan pria. \nDeskripsi: Sepatu ringan dengan bantalan empuk cocok untuk olahraga dan marathon.\n',
 'Nama Produk: Sepatu casual wanita. \nDeskripsi: Sepatu casual stylish dan nyaman untuk aktivitas sehari-hari.\n',
 'Nama Produk: Laptop gaming RTX 4060. \nDeskripsi: Laptop gaming performa tinggi dengan GPU RTX 4060 dan RAM 16GB.\n',
 'Nama Produk: Laptop kerja tipis ringan. \nDeskripsi: Laptop tipis dan ringan cocok untuk kerja kantoran dan presentasi.\n',
 'Nama Produk: Headphone noise cancelling. \nDeskripsi: Headphone dengan fitur noise cancelling aktif dan suara premium.\n']

In [ ]:
# metadata = semua kolom kecuali nama_produk dan deskripsi
metadatas = df.drop(columns=["nama_produk", "deskripsi"]).to_dict(orient="records")
metadatas

[{'id': 1,
  'kategori': 'sepatu',
  'brand': 'Nike',
  'harga': 950000,
  'rating': 4.7,
  'stok': True},
 {'id': 2,
  'kategori': 'sepatu',
  'brand': 'Adidas',
  'harga': 850000,
  'rating': 4.5,
  'stok': True},
 {'id': 3,
  'kategori': 'laptop',
  'brand': 'Asus',
  'harga': 18500000,
  'rating': 4.8,
  'stok': False},
 {'id': 4,
  'kategori': 'laptop',
  'brand': 'Lenovo',
  'harga': 12500000,
  'rating': 4.6,
  'stok': True},
 {'id': 5,
  'kategori': 'aksesoris',
  'brand': 'Sony',
  'harga': 2500000,
  'rating': 4.9,
  'stok': True}]

In [ ]:
# Buat embedding
embeddings = model.encode(documents).tolist()

In [ ]:
# Insert data ke Chroma
collection.add(
    documents=documents,
    embeddings=embeddings,
    metadatas=metadatas,
    ids=df["id"].astype(str).tolist()
)

In [ ]:
# cek collection
print(collection.count())
display(collection.get())

5


{'ids': ['1', '2', '3', '4', '5'],
 'embeddings': None,
 'documents': ['Nama Produk: Sepatu lari ringan pria. \nDeskripsi: Sepatu ringan dengan bantalan empuk cocok untuk olahraga dan marathon.\n',
  'Nama Produk: Sepatu casual wanita. \nDeskripsi: Sepatu casual stylish dan nyaman untuk aktivitas sehari-hari.\n',
  'Nama Produk: Laptop gaming RTX 4060. \nDeskripsi: Laptop gaming performa tinggi dengan GPU RTX 4060 dan RAM 16GB.\n',
  'Nama Produk: Laptop kerja tipis ringan. \nDeskripsi: Laptop tipis dan ringan cocok untuk kerja kantoran dan presentasi.\n',
  'Nama Produk: Headphone noise cancelling. \nDeskripsi: Headphone dengan fitur noise cancelling aktif dan suara premium.\n'],
 'uris': None,
 'included': ['metadatas', 'documents'],
 'data': None,
 'metadatas': [{'kategori': 'sepatu',
   'stok': True,
   'rating': 4.7,
   'id': 1,
   'harga': 950000,
   'brand': 'Nike'},
  {'id': 2,
   'rating': 4.5,
   'stok': True,
   'brand': 'Adidas',
   'harga': 850000,
   'kategori': 'sepatu'}

Selanjutnya kita akan coba melakukan pencarian dengan filtering, misalkan User ingin laptop dengan harga <= 13 juta dan stok tersedia. Kita akan menggunakan `where` di ChromaDB dan menggunakan beberapa operator.

Operator yang sering muncul atau digunakan :

Operator Perbandingan:

| Operator | Artinya                    |
| -------- | -------------------------- |
| `$eq`    | equal (=)                  |
| `$ne`    | not equal (!=)             |
| `$gt`    | greater than (>)           |
| `$gte`   | greater than or equal (>=) |
| `$lt`    | less than (<)              |
| `$lte`   | less than or equal (<=)    |


Operator Logika:

| Operator | Artinya                          |
| -------- | -------------------------------- |
| `$and`   | Semua kondisi harus terpenuhi    |
| `$or`    | Salah satu kondisi terpenuhi     |


In [ ]:
query = "laptop ringan dan cocok untuk kerja"
query_embedding = model.encode([query]).tolist()


results = collection.query(
    query_embeddings=query_embedding,
    n_results=3,
    where={
        "$and": [
            {"kategori": "laptop"},
            {"harga": {"$lte": 13000000}},
            {"stok": True}
        ]
    }
)

for doc, meta, dist in zip(
    results["documents"][0],
    results["metadatas"][0],
    results["distances"][0]
):
    print("Document:")
    print(doc)
    print('metadata:')
    print("Brand:", meta["brand"])
    print("Harga:", meta["harga"])
    print("Rating:", meta["rating"])
    print("-" * 40)

Document:
Nama Produk: Laptop kerja tipis ringan. 
Deskripsi: Laptop tipis dan ringan cocok untuk kerja kantoran dan presentasi.

metadata:
Brand: Lenovo
Harga: 12500000
Rating: 4.6
----------------------------------------


Sekarang, kita akan mencoba mencari `Sepatu lari ringan untuk olahraga` dan menambahkan filter threshold berdasarkan distance.

```
Ingat:
Distance mendekati 0 maka semakin mirip dengan query.
```

Kita ingin filter yang distancenya lebih kecil dari 0.5.

In [ ]:
import textwrap

# misalkan kita ingin mencari sepatu lari
query =  "Sepatu lari ringan untuk olahraga."
query_embedding = model.encode([query]).tolist()

results = collection.query(
    query_embeddings=query_embedding,
    n_results=3
)

threshold = 0.5

filtered = [
    (doc, dist)
    for doc, dist in zip(results["documents"][0], results["distances"][0])
    if dist < threshold
]

print("Hasil query dengan filter distance < 0.5:")
for doc, dist in filtered:
    print(f"Distance: {dist:.4f}")
    print("Document:")
    print(textwrap.indent(doc, "\t"))

Hasil query dengan filter distance < 0.5:
Distance: 0.3180
Document:
	Nama Produk: Sepatu lari ringan pria. 
	Deskripsi: Sepatu ringan dengan bantalan empuk cocok untuk olahraga dan marathon.

Distance: 0.4509
Document:
	Nama Produk: Sepatu casual wanita. 
	Deskripsi: Sepatu casual stylish dan nyaman untuk aktivitas sehari-hari.

Distance: 0.4877
Document:
	Nama Produk: Laptop kerja tipis ringan. 
	Deskripsi: Laptop tipis dan ringan cocok untuk kerja kantoran dan presentasi.



### Hybrid Search di ChromaDB

Chroma versi cloud menyedikan API atau dukungan khusus untuk pencarian hybrid. Namun, disini kita tidak akan gunakan versi cloud. Kita akan coba mengkombinasikan keyword search (BM25) dengan semantic search (dense retrival dengan all-MiniLM-L6-v2).

Pertama-tama kita akan mencoba pencarian menggunakan semantic search.

In [ ]:
# Dense Search via Chroma dengan model All-MiniLM-L6-v2
query =  "Sepatu lari ringan untuk olahraga."
query_embedding = model.encode([query])

dense_results = collection.query(
    query_embeddings=query_embedding.tolist(),
    n_results=len(documents)
)

dense_docs = dense_results["documents"][0]
dense_distances = dense_results["distances"][0]
dense_docs

['Nama Produk: Sepatu lari ringan pria. \nDeskripsi: Sepatu ringan dengan bantalan empuk cocok untuk olahraga dan marathon.\n',
 'Nama Produk: Sepatu casual wanita. \nDeskripsi: Sepatu casual stylish dan nyaman untuk aktivitas sehari-hari.\n',
 'Nama Produk: Laptop kerja tipis ringan. \nDeskripsi: Laptop tipis dan ringan cocok untuk kerja kantoran dan presentasi.\n',
 'Nama Produk: Headphone noise cancelling. \nDeskripsi: Headphone dengan fitur noise cancelling aktif dan suara premium.\n',
 'Nama Produk: Laptop gaming RTX 4060. \nDeskripsi: Laptop gaming performa tinggi dengan GPU RTX 4060 dan RAM 16GB.\n']

In [ ]:
dense_distances

[0.31802982091903687,
 0.4509098529815674,
 0.48773348331451416,
 0.6256129741668701,
 0.7317017912864685]

Selanjutnya, kita akan mencoba mengurangi skor dari distance dengan 1. Tujuannya agar skor yang tinggi berada di atas skor yang rendah.

In [ ]:
import numpy as np
dense_scores = 1 - np.array(dense_distances)
dense_scores

array([0.68197018, 0.54909015, 0.51226652, 0.37438703, 0.26829821])

Selanjutnya, kita akan mencoba pencarian menggunakan keyword search.

In [ ]:
from rank_bm25 import BM25Okapi

tokenized_docs = [doc.lower().split() for doc in documents]

print("Tokenized documents:")
for doc in tokenized_docs:
    print(doc)

Tokenized documents:
['nama', 'produk:', 'sepatu', 'lari', 'ringan', 'pria.', 'deskripsi:', 'sepatu', 'ringan', 'dengan', 'bantalan', 'empuk', 'cocok', 'untuk', 'olahraga', 'dan', 'marathon.']
['nama', 'produk:', 'sepatu', 'casual', 'wanita.', 'deskripsi:', 'sepatu', 'casual', 'stylish', 'dan', 'nyaman', 'untuk', 'aktivitas', 'sehari-hari.']
['nama', 'produk:', 'laptop', 'gaming', 'rtx', '4060.', 'deskripsi:', 'laptop', 'gaming', 'performa', 'tinggi', 'dengan', 'gpu', 'rtx', '4060', 'dan', 'ram', '16gb.']
['nama', 'produk:', 'laptop', 'kerja', 'tipis', 'ringan.', 'deskripsi:', 'laptop', 'tipis', 'dan', 'ringan', 'cocok', 'untuk', 'kerja', 'kantoran', 'dan', 'presentasi.']
['nama', 'produk:', 'headphone', 'noise', 'cancelling.', 'deskripsi:', 'headphone', 'dengan', 'fitur', 'noise', 'cancelling', 'aktif', 'dan', 'suara', 'premium.']


In [ ]:
bm25 = BM25Okapi(tokenized_docs)
print("IDF:", bm25.idf)
print("Average doc length:", bm25.avgdl)

IDF: {'nama': 0.16155784770889026, 'produk:': 0.16155784770889026, 'sepatu': 0.33647223662121295, 'lari': 1.0986122886681098, 'ringan': 0.33647223662121295, 'pria.': 1.0986122886681098, 'deskripsi:': 0.16155784770889026, 'dengan': 0.16155784770889026, 'bantalan': 1.0986122886681098, 'empuk': 1.0986122886681098, 'cocok': 0.33647223662121295, 'untuk': 0.16155784770889026, 'olahraga': 1.0986122886681098, 'dan': 0.16155784770889026, 'marathon.': 1.0986122886681098, 'casual': 1.0986122886681098, 'wanita.': 1.0986122886681098, 'stylish': 1.0986122886681098, 'nyaman': 1.0986122886681098, 'aktivitas': 1.0986122886681098, 'sehari-hari.': 1.0986122886681098, 'laptop': 0.33647223662121295, 'gaming': 1.0986122886681098, 'rtx': 1.0986122886681098, '4060.': 1.0986122886681098, 'performa': 1.0986122886681098, 'tinggi': 1.0986122886681098, 'gpu': 1.0986122886681098, '4060': 1.0986122886681098, 'ram': 1.0986122886681098, '16gb.': 1.0986122886681098, 'kerja': 1.0986122886681098, 'tipis': 1.0986122886681

In [ ]:
bm25_scores = bm25.get_scores(query.lower().split())

Lakukan normalisasi skor agar berada dalam rentang 0-1

In [ ]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

# normalisasi skor agar berada dalam rentang 0-1
dense_scores_norm = scaler.fit_transform(dense_scores.reshape(-1,1)).flatten()
bm25_scores_norm = scaler.fit_transform(bm25_scores.reshape(-1,1)).flatten()

Setelah melakukan normalisasi, kita akan mencoba mengkombinasikan skor dari semantic search dan keyword search kemudian melakukan rangking. Skor terbesar akan menjadi skor terbaik atau yang paling relevan dengan query yang diberikan. Untuk perhitungan ini kita akan menggunakan `Weighted Linear Combination` atau `Linear Score Fusion`.

In [ ]:
# Hybrid Fusion
alpha = 0.6

hybrid_scores = alpha * dense_scores_norm + (1 - alpha) * bm25_scores_norm

results = list(zip(documents, hybrid_scores))
results = sorted(results, key=lambda x: x[1], reverse=True)

print("=== HYBRID SEARCH ===\n")
for doc, score in results:
    print(f"{score:.4f} - {doc}")

=== HYBRID SEARCH ===

1.0000 - Nama Produk: Sepatu lari ringan pria. 
Deskripsi: Sepatu ringan dengan bantalan empuk cocok untuk olahraga dan marathon.

0.5311 - Nama Produk: Sepatu casual wanita. 
Deskripsi: Sepatu casual stylish dan nyaman untuk aktivitas sehari-hari.

0.3539 - Nama Produk: Laptop gaming RTX 4060. 
Deskripsi: Laptop gaming performa tinggi dengan GPU RTX 4060 dan RAM 16GB.

0.2433 - Nama Produk: Laptop kerja tipis ringan. 
Deskripsi: Laptop tipis dan ringan cocok untuk kerja kantoran dan presentasi.

0.0000 - Nama Produk: Headphone noise cancelling. 
Deskripsi: Headphone dengan fitur noise cancelling aktif dan suara premium.



Dari hasil diatas kita dapat melihat selisih skor yang cukup besar antara antara produk sepatu dan laptop sehingga kita dapat dengan mudah menentukan produk mana yang lebih relevan, bahkan skor dari sepatu lari ringan untuk olahraga adalah 1. Berbeda dengan sebelum kita melakukan hybrid search, produk laptop tetap cenderung memiliki nilai yang berdekatan dengan sepatu.

## Pinecone

Selanjutnya kita akan gunakan hybrid search di Pinecone. Data yang digunakan tetap dari dataframe yang telah kita buat sebelumnya.

In [ ]:
df

,id,nama_produk,deskripsi,kategori,brand,harga,rating,stok
0,1,Sepatu lari ringan pria,Sepatu ringan dengan bantalan empuk cocok untu...,sepatu,Nike,950000,4.7,True
1,2,Sepatu casual wanita,Sepatu casual stylish dan nyaman untuk aktivit...,sepatu,Adidas,850000,4.5,True
2,3,Laptop gaming RTX 4060,Laptop gaming performa tinggi dengan GPU RTX 4...,laptop,Asus,18500000,4.8,False
3,4,Laptop kerja tipis ringan,Laptop tipis dan ringan cocok untuk kerja kant...,laptop,Lenovo,12500000,4.6,True
4,5,Headphone noise cancelling,Headphone dengan fitur noise cancelling aktif ...,aksesoris,Sony,2500000,4.9,True


Pertama-tama kita coba install terlebih dahulu library `pinecone_text`.  `pinecone_text` adalah library tambahan dari Pinecone yang dulu dipakai untuk membuat sparse vector (keyword-based) agar bisa melakukan hybrid search (Dense + BM25) di Pinecone.

In [ ]:
# !pip install pinecone-text

In [ ]:
from pinecone import Pinecone, ServerlessSpec
from pinecone_text.sparse import BM25Encoder
from sentence_transformers import SentenceTransformer
import numpy as np

Selanjutnya kita akan lakukan inisialisasi Pinecone dan membuat index baru.

In [ ]:
# Inisialisasi Pinecone client
PINECONE_API_KEY = "your-pinecone-api-key"
pc = Pinecone(api_key=PINECONE_API_KEY)
index_name = "hs-products"

In [ ]:
# hapus index pinecone
# pc.delete_index('latihan-semantic-search')
# pc.delete_index(index_name)

Di pinecone untuk melakukan hybrid search, karena menggunakan sparse value maka kita perlu mengatur metric menjadi `dotproduct`. Jika tidak diatur ke dotproduct maka akan error saat melakukan query atau search.

In [ ]:
# buat index baru jika belum ada
if index_name not in pc.list_indexes().names():
    pc.create_index(
        name=index_name,
        dimension=384,  # all-MiniLM-L6-v2 = 384
        metric="dotproduct",  # wajib dotproduct , karena gunakan sparse
        spec=ServerlessSpec(
            cloud="aws",
            region="us-east-1"
        )
    )

index = pc.Index(index_name)

Setelah berhasil membuat index baru, kita akan mencoba melakukan inisialisasi model dense (semantic) dan sparse (keyword).

In [ ]:
# Gabungkan nama + deskripsi jadi 1 document
documents = [
f"""Nama Produk: {row.nama_produk}.
Deskripsi: {row.deskripsi}
"""
for _, row in df.iterrows()
]

In [ ]:
#inisialisasi dense dan sparse
# dense embedding dengan SentenceTransformer
model = SentenceTransformer("all-MiniLM-L6-v2")

# sparse dengan BM25
bm25=BM25Encoder()
# Fit BM25 pada korpus teks
bm25.fit(documents)


Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


  0%|          | 0/5 [00:00<?, ?it/s]

Selanjutnya kita akan melakukan encoding untuk membuat dense vector dan sparse vector.

In [ ]:
# Generate Dense Vectors
dense_vectors = model.encode(documents, show_progress_bar=True).tolist()

# Generate Sparse Vectors
sparse_vectors = bm25.encode_documents(documents)

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

Selanjutnya kita akan melakukan upsert ke Pinecone. Karena pinecone tidak menyimpan deksripsi dan nama_produk secara langsung, maka kita perlu menyimpannya sebagai metadata agar mempermudah dalam diakses ketika melakukan query.

In [ ]:
data_to_upsert = []

for i, row in df.iterrows():
    data_to_upsert.append({
        "id": str(row["id"]),
        "values": dense_vectors[i],
        "sparse_values": sparse_vectors[i],
        "metadata": {
            "nama_produk": row["nama_produk"],
            "deskripsi": row["deskripsi"], # Penting untuk level production
            "kategori": row["kategori"],
            "harga": row["harga"],
            "brand": row["brand"],
            "stok": row["stok"],
        }
    })

# Upsert ke Pinecone
index.upsert(vectors=data_to_upsert)

UpsertResponse(upserted_count=5, _response_info={'raw_headers': {'date': 'Tue, 03 Mar 2026 09:55:48 GMT', 'content-type': 'application/json', 'content-length': '19', 'connection': 'keep-alive', 'x-pinecone-request-lsn': '1', 'x-pinecone-request-logical-size': '9187', 'x-pinecone-request-latency-ms': '645', 'x-envoy-upstream-service-time': '252', 'x-pinecone-response-duration-ms': '647', 'grpc-status': '0', 'server': 'envoy'}})

In [ ]:
# cek statistik dari index
index.describe_index_stats()

{'_response_info': {'raw_headers': {'connection': 'keep-alive',
                                    'content-length': '185',
                                    'content-type': 'application/json',
                                    'date': 'Tue, 03 Mar 2026 09:55:49 GMT',
                                    'grpc-status': '0',
                                    'server': 'envoy',
                                    'x-envoy-upstream-service-time': '38',
                                    'x-pinecone-request-latency-ms': '37',
                                    'x-pinecone-response-duration-ms': '39'}},
 'dimension': 384,
 'index_fullness': 0.0,
 'memoryFullness': 0.0,
 'metric': 'dotproduct',
 'namespaces': {'__default__': {'vector_count': 5}},
 'storageFullness': 0.0,
 'total_vector_count': 5,
 'vector_type': 'dense'}

In [ ]:
query='Sepatu lari ringan untuk olahraga.'

dense_query = model.encode(query).tolist()
sparse_query = bm25.encode_queries(query)

In [ ]:
dense_query

[-0.029078368097543716,
 0.10753735899925232,
 -0.12139499187469482,
 -0.014624102041125298,
 -0.06753017008304596,
 -0.04958808049559593,
 0.04677962884306908,
 -0.008858450688421726,
 0.08037101477384567,
 0.003526863409206271,
 0.06548496335744858,
 -0.030913665890693665,
 -0.006136151496320963,
 -0.05444134771823883,
 0.037050750106573105,
 -0.037014175206422806,
 -0.053363971412181854,
 0.0431753508746624,
 -0.12244858592748642,
 -0.055082593113183975,
 0.02431715838611126,
 0.034960586577653885,
 0.00806413497775793,
 0.046638041734695435,
 0.03597087785601616,
 0.039972539991140366,
 0.023941151797771454,
 0.010036404244601727,
 -0.011377225629985332,
 -0.024867163971066475,
 0.08061600476503372,
 0.06272410601377487,
 0.011718212626874447,
 -0.029181551188230515,
 -0.02029193378984928,
 0.03090813383460045,
 -0.10032862424850464,
 -0.0386866070330143,
 -0.021590331569314003,
 0.024561384692788124,
 0.007316868286579847,
 0.02014152891933918,
 -0.01881265453994274,
 -0.117180131

In [ ]:
sparse_query

{'indices': [2225900446, 2201728864, 2825008493, 813961998, 3019017155],
 'values': [0.17293132093117414,
  0.2738346954484004,
  0.17293132093117414,
  0.1064679672408509,
  0.2738346954484004]}

In [ ]:
# Query ke Pinecone
results = index.query(
    vector=dense_query,
    sparse_vector=sparse_query,
    top_k=5,
    include_metadata=True
)

# ambil hasil query
for match in results["matches"]:
   print('score:', match["score"])
   print("nama produk:", match["metadata"]["nama_produk"])
   print('='*40)

score: 1.18753338
nama produk: Sepatu lari ringan pria
score: 0.713452101
nama produk: Sepatu casual wanita
score: 0.666575611
nama produk: Laptop kerja tipis ringan
score: 0.37465
nama produk: Headphone noise cancelling
score: 0.268323898
nama produk: Laptop gaming RTX 4060


Agar lebih rapi dan lebih akurat sesuai dengan kebutuhan, kita dapat membuat fungsi dan menerapkan Weighted Hybrid Search atau yang sering disebut sebagai Alpha Tuning yang berfungsi untuk mengontrol perbandingan antara hasil dari semantic search dan keyword search.

- Jika Alpha = 1.0: Anda melakukan pencarian 100% Semantic. Kata kunci spesifik seringkali terabaikan demi makna umum.
- Jika Alpha = 0.0: Anda melakukan pencarian 100% Keyword (BM25). Hanya dokumen dengan kata yang sama persis yang akan muncul.
- Jika Alpha = 0.5 (Ideal): Keduanya memiliki bobot yang sama. Ini adalah titik awal terbaik untuk General Purpose Search.

Referensi: [Hybrid Search - Pinecone](https://docs.pinecone.io/guides/search/hybrid-search)

In [ ]:
def hybrid_query(query, top_k=3, alpha=0.5, filter_dict=None):
    # Encode Query
    dense_q = model.encode(query).tolist()
    sparse_q = bm25.encode_queries(query)

    # Hybrid Scaling: (Dense * alpha) & (Sparse * (1 - alpha))
    dense = [v * alpha for v in dense_q]
    sparse = {
        'indices': sparse_q['indices'],
        'values': [v * (1 - alpha) for v in sparse_q['values']]
    }

    # Query ke Pinecone
    result = index.query(
        vector=dense,
        sparse_vector=sparse,
        top_k=top_k,
        include_metadata=True,
        filter=filter_dict
    )

    return result['matches']

In [ ]:
query='Sepatu lari ringan untuk olahraga.'
results = hybrid_query(query, top_k=5, alpha=0.5)

for match in results:
    print('score:', match["score"])
    print("nama produk:", match["metadata"]["nama_produk"])
    print("deskripsi:", match["metadata"]["deskripsi"])
    print('='*40)

score: 0.593766689
nama produk: Sepatu lari ringan pria
deskripsi: Sepatu ringan dengan bantalan empuk cocok untuk olahraga dan marathon.
score: 0.35672605
nama produk: Sepatu casual wanita
deskripsi: Sepatu casual stylish dan nyaman untuk aktivitas sehari-hari.
score: 0.333287805
nama produk: Laptop kerja tipis ringan
deskripsi: Laptop tipis dan ringan cocok untuk kerja kantoran dan presentasi.
score: 0.187325
nama produk: Headphone noise cancelling
deskripsi: Headphone dengan fitur noise cancelling aktif dan suara premium.
score: 0.134161949
nama produk: Laptop gaming RTX 4060
deskripsi: Laptop gaming performa tinggi dengan GPU RTX 4060 dan RAM 16GB.


Sekarang skornya berada pada rentang 0-1, skor yang rendah menandakan kemiripan yang rendah, sedangkan skor yang tinggi menandakan kemiripan yang tinggi.

Selanjutnya, kita dapat melakukan pencarian dengan menambahkan filter jika ingin lebih spesifik.

In [ ]:
query='Sepatu lari ringan untuk olahraga.'
results = hybrid_query(query, top_k=5, alpha=0.5, filter_dict={"kategori": "sepatu"})

for match in results:
    print('score:', match["score"])
    print("nama produk:", match["metadata"]["nama_produk"])
    print("deskripsi:", match["metadata"]["deskripsi"])
    print('='*40)

score: 0.593766689
nama produk: Sepatu lari ringan pria
deskripsi: Sepatu ringan dengan bantalan empuk cocok untuk olahraga dan marathon.
score: 0.35672605
nama produk: Sepatu casual wanita
deskripsi: Sepatu casual stylish dan nyaman untuk aktivitas sehari-hari.


In [ ]:
query='Sepatu lari ringan untuk olahraga.'
results = hybrid_query(query,
                       top_k=5,
                       alpha=0.5,
                       filter_dict={
                        "$and": [
                            {"kategori": "sepatu"},
                            {"brand": "Nike"},
                            {"harga": {"$lt": 1000000}}
                            ]
                        }
)

for match in results:
    print('score:', match["score"])
    print("nama produk:", match["metadata"]["nama_produk"])
    print("deskripsi:", match["metadata"]["deskripsi"])
    print('='*40)

score: 0.593766689
nama produk: Sepatu lari ringan pria
deskripsi: Sepatu ringan dengan bantalan empuk cocok untuk olahraga dan marathon.


Selain cara tersebut, kita juga dapat menggunakan langchain untuk melakukan Hybrid Search tanpa membuat fungsi hybrid_query.

sebelum melakukan hybrid search, kita akan coba hapus semua data di index dan memasukan kembali dokumen dan metadata dengan metode `.add_texts()`

In [ ]:
# hapus semua data di index
index.delete(delete_all=True)

{'_response_info': {'raw_headers': {'date': 'Tue, 03 Mar 2026 10:00:07 GMT',
   'content-type': 'application/json',
   'content-length': '2',
   'connection': 'keep-alive',
   'x-pinecone-request-latency-ms': '172',
   'x-envoy-upstream-service-time': '176',
   'x-pinecone-response-duration-ms': '178',
   'grpc-status': '0',
   'server': 'envoy'}}}

In [ ]:
# cek index apakah total vector sudah 0
index.describe_index_stats()

{'_response_info': {'raw_headers': {'connection': 'keep-alive',
                                    'content-length': '154',
                                    'content-type': 'application/json',
                                    'date': 'Tue, 03 Mar 2026 10:00:08 GMT',
                                    'grpc-status': '0',
                                    'server': 'envoy',
                                    'x-envoy-upstream-service-time': '1',
                                    'x-pinecone-request-latency-ms': '1',
                                    'x-pinecone-response-duration-ms': '2'}},
 'dimension': 384,
 'index_fullness': 0.0,
 'memoryFullness': 0.0,
 'metric': 'dotproduct',
 'namespaces': {},
 'storageFullness': 0.0,
 'total_vector_count': 0,
 'vector_type': 'dense'}

Selanjutnya kita akan melakukan inisialisasi embeddings dan sparse encoder.

In [ ]:
from langchain_community.embeddings import HuggingFaceEmbeddings
from pinecone_text.sparse import BM25Encoder
from langchain_community.retrievers import PineconeHybridSearchRetriever
# Dense model
embeddings = HuggingFaceEmbeddings(
    model_name="sentence-transformers/all-MiniLM-L6-v2"
)

# Sparse encoder
bm25_encoder = BM25Encoder()
bm25_encoder.fit(documents)

C:\Users\sardi\AppData\Local\Temp\ipykernel_12904\3391876322.py:5: LangChainDeprecationWarning: The class `HuggingFaceEmbeddings` was deprecated in LangChain 0.2.2 and will be removed in 1.0. An updated version of the class exists in the `langchain-huggingface package and should be used instead. To use it run `pip install -U `langchain-huggingface` and import as `from `langchain_huggingface import HuggingFaceEmbeddings``.
  embeddings = HuggingFaceEmbeddings(


Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


  0%|          | 0/5 [00:00<?, ?it/s]

Setelah inisialisasi embeddings dan sparse encoder, kita akan inisialisasi retriever hybrid menggunakan method `PineconeHybridSearchRetriever`. Method ini berguna sebagai wrapper untuk melakukan hybrid search.

In [ ]:
index=pc.Index(index_name)

# Inisialisasi retriever hybrid
retriever = PineconeHybridSearchRetriever(
    embeddings=embeddings, sparse_encoder=bm25_encoder,
    index=index,
    top_k=3 # jumlah hasil yang ingin ditampilkan
)

Selanjutnya kita akan menggunakan method `add_texts` untuk menambahkan dokumen dan metadata ke retriever atau Pinecone. Ini mirip seperti upsert.

In [ ]:
df["id"].astype(str).tolist()

['1', '2', '3', '4', '5']

In [ ]:
documents = [
f"""Nama Produk: {row.nama_produk}.
Deskripsi: {row.deskripsi}
"""
for _, row in df.iterrows()
]

metadatas= df.to_dict(orient="records")

In [ ]:
# tambahkan dokumen dan metadata ke retriever (Pinecone)
retriever.add_texts(documents, metadatas=metadatas, ids=df["id"].astype(str).tolist())


  0%|          | 0/1 [00:00<?, ?it/s]

Kita cek kembali apakah index sudah terisi.

In [ ]:
# cek statistik dari index
index.describe_index_stats()

{'_response_info': {'raw_headers': {'connection': 'keep-alive',
                                    'content-length': '185',
                                    'content-type': 'application/json',
                                    'date': 'Tue, 03 Mar 2026 10:01:03 GMT',
                                    'grpc-status': '0',
                                    'server': 'envoy',
                                    'x-envoy-upstream-service-time': '4',
                                    'x-pinecone-request-latency-ms': '3',
                                    'x-pinecone-response-duration-ms': '5'}},
 'dimension': 384,
 'index_fullness': 0.0,
 'memoryFullness': 0.0,
 'metric': 'dotproduct',
 'namespaces': {'__default__': {'vector_count': 5}},
 'storageFullness': 0.0,
 'total_vector_count': 5,
 'vector_type': 'dense'}

In [ ]:
response_1 = index.fetch(ids=["1"])
print(response_1)

FetchResponse(namespace='', vectors={'1': Vector(id='1', values=[-0.0532917976, 0.0429131575, -0.0656574368, -0.0760856867, -0.0487533659, 0.0343126543, 0.0460214391, 0.0674621761, 0.0274529029, -0.016123414, 0.102737151, -0.0293453634, -0.0348456353, -0.0567745119, 0.0426142514, -0.0887916, 0.023383446, 0.0510237627, -0.0791495219, -0.0900121257, 0.0579649061, -0.0584304258, -0.0262051839, 0.0403739028, -0.0248811785, 0.0453637466, 0.0380747057, 0.0208344553, 0.0202395674, -0.0514605716, 0.00605571596, 0.0285921022, 0.0658970922, -0.0126497671, 0.0572818443, 0.07339295, -0.0698067322, -0.0412148088, 0.0250757951, -0.0207956154, 0.0494992211, -0.0542971939, -0.0855429, -0.14061904, 0.0680793747, -0.0217831302, -0.0484087244, -0.00632424885, 0.0488111265, 0.00204287237, -0.0682289302, 0.0541386455, -0.0353372395, 0.00637431769, -0.0373554, -0.0710372776, 0.0318186134, -0.0167842042, -0.00692120567, 0.0193624832, -0.0554708801, -0.0108001065, -0.0460043214, 0.0378791355, 0.0212449413, -0

In [ ]:
# tampilkan metadata dari vector dengan id "1"
print("Metadata for ID 1:")
print(response_1["vectors"]["1"]["metadata"])

Metadata for ID 1:
{'brand': 'Nike', 'context': 'Nama Produk: Sepatu lari ringan pria. \nDeskripsi: Sepatu ringan dengan bantalan empuk cocok untuk olahraga dan marathon.\n', 'harga': 950000, 'id': 1, 'kategori': 'sepatu', 'rating': 4.7, 'stok': True}


Setelah index terisi dengan vektor dan metadata, kita dapat melakukan pencarian dengan retriever.

In [ ]:
query= "Sepatu lari ringan untuk olahraga."
results = retriever.invoke(query, filter={"kategori": "sepatu"})

# gunakan sorted() untuk memastikan agar hasilnya berurutan berdasarkan skor tertinggi
sorted_results = sorted(
    results,
    key=lambda x: x.metadata.get('score', 0),
    reverse=True # urutkan dari yang paling relevan (score tertinggi)
)


In [ ]:
for doc in sorted_results:
    # Cek apakah score ada di metadata (beberapa versi LangChain menaruhnya di sini)
    score = doc.metadata.get("score", "0")
    print(f"Score: {score}")
    print(f"Content: {doc.page_content}")
    print("-" * 20)

Score: 0.593766689
Content: Nama Produk: Sepatu lari ringan pria. 
Deskripsi: Sepatu ringan dengan bantalan empuk cocok untuk olahraga dan marathon.

--------------------
Score: 0.35672605
Content: Nama Produk: Sepatu casual wanita. 
Deskripsi: Sepatu casual stylish dan nyaman untuk aktivitas sehari-hari.

--------------------


# Tugas

Setelah Anda mempelari cara untuk membaca dokumen PDF, Chunking, membuat metadata dan melakukan Hybrid Search. Sekarang coba gunakan data atau dokumen PDF yang ada pada sesi ini, untuk mencari konteks atau teks yang cocok dengan input atau query berikut:
- Berapa jumlah BTS di PT NTD.
- Berikan Langkah registrasi prabayar.
- Layanan apa saja yang ada di PT NTD.
- Berikan daftar paket Internet Paling Murah.

Lakukan pencarian dengan Hybrid Search dan Vector Database yang digunakan adalah Pinecone.

Hint: Untuk melakukan Embedding, gunakan hasil dari Chunking untuk dijadikan sebagai dokumen, misalkan :
```
embeddings = model.encode(chunk).tolist()
```